axlearn/common/checkpointer_orbax_emergency.py

```diff
@@ -0,0 +1,814 @@
+        """
+        cfg: OrbaxEmergencyCheckpointer.Config = self.config
+        if self._tensor_manager is not None:
+            return self._tensor_manager
+
+        save_policy = cfg.save_policy.instantiate()
+        local_save_policy = cfg.local_save_policy.instantiate()
+
+        def _orbax_save_fn(
+            step: int, last_saved_step: Optional[int], wrapped_save_policy: CheckpointPolicy
+        ) -> bool:
+            del last_saved_step
+            return wrapped_save_policy(step=step, evaler_summaries=self._eval_summaries)
+
+        # For meaning of these options, refer to
+        # https://github.com/google/orbax/blob/95be2c021bc8cbf4badd83a053ff57b7a9f9b314/checkpoint/orbax/checkpoint/experimental/emergency/checkpoint_manager.py#L277
+        self._tensor_manager = oecp.CheckpointManager(
+            self._local_dir,
+            persistent_directory=os.path.join(cfg.dir, self._TENSORS_PREFIX),
+            global_mesh=thread_resources.env.physical_mesh,
+            abstract_state=self._get_abstract_state(state_with_tensors),
+            options=oecp.CheckpointManagerOptions(
+                local=oecp.LocalCheckpointOptions(
+                    should_save_fn=functools.partial(
+                        _orbax_save_fn, wrapped_save_policy=local_save_policy
+                    ),
+                    max_to_keep=cfg.local_keep_last_n,
+                ),
+                persistent=oecp.PersistentCheckpointOptions(
+                    should_save_fn=functools.partial(
+                        _orbax_save_fn, wrapped_save_policy=save_policy
+                    ),
+                    max_to_keep=cfg.keep_last_n,
+                ),
+                replica_axis_index=cfg.replica_axis_index,
+                async_options=oecp.checkpoint_manager.AsyncOptions(
+                    timeout_secs=cfg.async_timeout_secs
+                ),
+                step_name_format=self._name_format,
+                cleanup_tmp_directories=True,
+                enable_async_checkpointing=True,
+            ),
+        )
+        return self._tensor_manager
+
+    def save(
+        self, *, step: int, state: Nested[Tensor], evaler_summaries: Optional[Dict[str, Any]] = None
+    ):
+        """See `BaseCheckpointer.save` for details."""
+        assert self._eval_summaries is None, self._eval_summaries
+        self._eval_summaries = copy.deepcopy(evaler_summaries or {})
+        self._reached_preemption = self._tensor_manager.reached_preemption(step)
+
+        start_t = time.perf_counter()
+        state_with_tensors = jax.tree.map(
+            lambda x: x if isinstance(x, (Tensor, TensorSpec)) else None, state
+        )
+        # Note that save() waits for prior serialization to finish.
+        self._non_tensor_manager.save(step=step, state=state)
+        self._get_tensor_manager(state_with_tensors).save(
+            step=step, args=ocp.args.PyTreeSave(item=state_with_tensors)
+        )
```

Comment:
How do we mark the completion of a checkpoint? It should happen only when both tensor and non-tensor states are saved. How is this ensured?

Please add a comment.

Reply:
There's no special marker for completion of both. There are only markers for each of them individually. So, during restore, we look for both of them only load a specific step when both marker exists.

Reply:
Thanks. Can you add a comment here and point to where "we look for both of them only load a specific step when both marker exists"?

Do we have testing for incomplete checkpoints?

Reply:
I can add a testcase.